src/Conversion/ONNXToKrnl/PerfectHash.cpp

```diff
@@ -0,0 +1,211 @@
+  static SmallVector<uint32_t> range(uint32_t min, uint32_t max) {
+    SmallVector<uint32_t> range(max - min);
+    std::iota(range.begin(), range.end(), min);
+    return range;
+  }
+
+  // Generate the integers [min, min+step, ...].
+  static SmallVector<uint32_t> range(int32_t min, int32_t max, int32_t step) {
+    SmallVector<uint32_t> range;
+    int32_t nElems = (max - min) / step;
+    if (nElems < 1)
+      return range;
+
+    range.resize(nElems);
+    int32_t num = min;
+    std::generate_n(range.begin(), nElems, [&num, step]() {
+      int32_t res = num;
+      num += step;
+      return res;
+    });
+    return range;
+  }
+
+  template <typename T>
+  static void print(const llvm::SmallVectorImpl<T> &V,
+      const llvm::StringRef name, llvm::raw_ostream &os) {
+    os << name << ": [ ";
+    for (const T &elem : V)
+      os << elem << ", ";
+    os << "]\n";
+  }
+
+  template <typename KeyType, typename ValueType>
+  static void print(const std::map<KeyType, ValueType> &M,
+      const llvm::StringRef name, llvm::raw_ostream &os) {
+    os << name << " : {";
+    for (const auto &entry : M)
+      os << "'" << entry.first << "': " << entry.second << ", ";
+    os << "}\n";
+  }
+};
+
+template <typename KeyTy, typename ValueTy>
+PerfectHash<KeyTy, ValueTy>::PerfectHash(const std::map<KeyTy, ValueTy> &dict)
+    : dict(dict) {
+  assert(!dict.empty() && "Dictionary should not be empty");
+  size_t size = dict.size();
+  G.resize(size, 0);
+  V.resize(size, -1);
+  createPerfectHash();
+}
+
+// Note: KeyTy is expected to be either a char* or a int64_t.
+// TODO: add a trait to ensure template cannot be instantiated with an
+// unexpected type.
+template <typename KeyTy, typename ValueTy>
+void PerfectHash<KeyTy, ValueTy>::createPerfectHash() {
+  LLVM_DEBUG({ Utilities::print(dict, "dict", dbgs()); });
+
+  // Step 1: place all of the keys into buckets.
+  size_t size = dict.size();
+  SmallVector<KeyTy> keys = Utilities::extractKeys<KeyTy, ValueTy>(dict);
+  LLVM_DEBUG({ Utilities::print(keys, "keys", dbgs()); });
+
+  SmallVector<SmallVector<KeyTy>> buckets(size);
+  for (const KeyTy &key : keys) {
+    buckets[Utilities::hash(0, key) % size].push_back(key);
+  }
+
+  // Step 2: Sort the buckets and process the ones with the most items first.
+  llvm::sort(buckets, [](const llvm::SmallVectorImpl<KeyTy> &v1,
+                          const llvm::SmallVectorImpl<KeyTy> &v2) {
+    return v1.size() > v2.size();
+  });
+
+  uint32_t biMax = 0;
+  for (uint32_t bi : Utilities::range(size)) {
+    LLVM_DEBUG(dbgs() << "bi=" << bi << "\n");
+    biMax = bi;
+    SmallVector<KeyTy> &bucket = buckets[bi];
+    if (bucket.size() <= 1)
+      break;
+
+    int32_t hval = 1;
+    size_t item = 0;
+    SmallVector<uint32_t> slots;
+
+    // Repeatedly try different hash values until we find a hash function that
+    // places all items in the bucket into free slots.
+    while (item < bucket.size()) {
```

Comment:
Will you have a escape if it repeat for ever? E.g. after iterating for X, escape and increase the size of the hash?

Reply:
The issue of potentially ending with an infinite loop here will be tracked via a separate PR. I have open a work item to track (https://github.com/onnx/onnx-mlir/issues/1010)